lib/Serializer.ts

```diff
@@ -0,0 +1,120 @@
+import { ObjectType, ModelType } from "./General";
+import { Document } from "./Document";
+
+const defaultSerializer = {
+	"modify": (serialized: ObjectType, original: ObjectType): ObjectType => {
+		return {...original};
+	}
+};
+
+const includeHandler = (document: ModelType<Document>, includeRules: string[], serialized: ObjectType): ObjectType => {
+	return includeRules.reduce((serialized: ObjectType, key: string) => {
+		if (Object.keys(document).includes(key)) {
+			serialized[key] = document[key];
+		}
+
+		return serialized;
+	}, serialized);
+};
+
+const excludeHandler = (document: ModelType<Document>, excludeRules: string[], serialized: ObjectType): ObjectType => {
+	return excludeRules.reduce((serialized: ObjectType, key: string) => {
+		if (Object.keys(serialized).includes(key)) {
+			delete serialized[key];
+		}
+
+		return serialized;
+	}, serialized);
+};
+
+const validateName = (name: string): void => {
+	if (!name || typeof name !== "string") {
+		throw new Error("Field name is required and should be of type string");
+	}
+};
+
+const validateOptions = (options): void => {
+	if (!options || !(Array.isArray(options) || typeof options === "object")) {
+		throw new Error("Field options is required and should be an object or array");
+	}
+};
+
+const cleanAndValidateDocumentsArray = (documentsArray: ModelType<Document>[]): ModelType<Document>[] => {
+	if (!documentsArray || !Array.isArray(documentsArray)) {
+		throw new Error("documentsArray must be an array of document objects");
+	}
+
+	return documentsArray.filter((doc) => typeof doc.serialize === "function");
+};
+
+export class Serializer {
+	#serializers: any;
+	#defaultSerializer: string;
+
+	constructor() {
+		this.#serializers = {_default: defaultSerializer};
+		this.#defaultSerializer = "_default";
+	}
+
+	add(name: string, options): void {
+		validateName(name);
+		validateOptions(options);
+		this.#serializers[name] = options;
+	}
+
+	setDefault(name: string): void {
+		validateName(name);
+		if (Object.keys(this.#serializers).includes(name)) {
+			this.#defaultSerializer = name;
+		}
+	}
+
+	remove(name: string): void {
+		validateName(name);
+		if (Object.keys(this.#serializers).includes(name)) {
+			delete this.#serializers[name];
+		}
+		if (this.#defaultSerializer === name) {
+			this.#defaultSerializer = "_default";
+		}
+	}
+
+	_serializeMany(documentsArray: ModelType<Document>[] = [], nameOrOptions): ObjectType[] {
```

Comment:
Wanna figure out if there is a way to use private methods for this and `_serialize`. And make it only accessible through Dynamoose. Since it won't be accessed outside of this library.

Reply:
I can't figure out a way to do this quickly, so gonna leave it as is. It'll be a project for another time.